lib/services/profiles-api-service.js

```diff
@@ -102,41 +102,80 @@ function profileApiServiceFactory(
 
                 });
             } else {
-                return self.createNodeAndRunDiscovery(macAddresses);
+                return self.createNodeAndRunDiscovery(macAddresses, options);
             }
         });
     };
 
-    ProfileApiService.prototype.runDiscovery = function(node) {
+    ProfileApiService.prototype.runDiscovery = function(node, options) {
         var self = this;
+        var configuration;
 
+        if (node.type === 'switch') {
+            configuration = self.getSwitchDiscoveryConfiguration(node, options.switchVendor);
+        } else {
+            configuration = {
+                name: 'Graph.SKU.Discovery',
+                options: {
+                    defaults: {
+                        graphOptions: {
+                            target: node.id
+                        },
+                        nodeId: node.id
+                    }
+                }
+            };
+        }
+
+        // The nested workflow holds the lock against the nodeId in this case,
+        // so don't add it as a target to the outer workflow context
+        return workflowApiService.createAndRunGraph(configuration, null)
+        .then(function() {
+            return self.waitForDiscoveryStart(node.id);
+        })
+        .then(function() {
+            return node;
+        });
+    };
+
+    ProfileApiService.prototype.getSwitchDiscoveryConfiguration = function(node, vendor) {
         var configuration = {
-            name: 'Graph.SKU.Discovery',
+            name: 'Graph.SKU.Switch.Discovery.Active',
             options: {
                 defaults: {
                     graphOptions: {
                         target: node.id
                     },
                     nodeId: node.id
+                },
+                'vendor-discovery-graph': {
+                    graphName: null
                 }
             }
         };
 
-        return workflowApiService.createAndRunGraph(configuration)
-        .then(function() {
-            return self.waitForDiscoveryStart(node.id);
-        })
-        .then(function() {
-            return node;
-        });
+        vendor = vendor.toLowerCase();
+
+        if (vendor === 'cisco') {
```

Comment:
@jlongever Like, move the script declarations out of the workflow and into a big python script? If I'm understanding correctly, I think inverting the control like this would probably mean we'd have to reinvent a few wheels that we already have?

This makes me think of another approach though, which could avoid the issue of hardcoding vendor types. What if we enforce a naming convention for cataloging scripts (e.g. `catalog_<vendor>_switch_<catalog-type>.py`), and then have a single cataloging workflow (e.g. catalog vendor task, catalog snmp task, etc.) that takes a vendor type as an option, and updates the downloadUrl, like (rough graph pseudo-definition):

```
// graph definition
{
    <stuff>,
    options: {
        defaults: {
            switchvendor: <set dynamically based on API params>
        }
    },
    tasks: [
        {
            label: "catalog-version",
            <stuff>,
            options: {
                switchvendor: null,
                downloadUrl: "{{ api.server }}/templates/catalog_{{ options.switchvendor }}_switch_version.py"
        },
        {
            label: "catalog-snmp",
            <stuff>,
            options: {
                switchvendor: null,
                downloadUrl: "{{ api.server }}/templates/catalog_{{ options.switchvendor }}_switch_snmp.py"
        }
    ]
}
```

This way, vendor types aren't hardcoded, just passed through dynamically, and we have one cataloging workflow definition to rule them all, rather than one per vendor.